llvm/test/Transforms/LoopVectorize/RISCV/vectorize-force-tail-with-evl-reduction.ll

```diff
@@ -0,0 +1,2253 @@
+; IF-EVL:       vector.ph:
+; IF-EVL-NEXT:    [[TMP4:%.*]] = call i64 @llvm.vscale.i64()
+; IF-EVL-NEXT:    [[TMP5:%.*]] = mul i64 [[TMP4]], 4
+; IF-EVL-NEXT:    [[TMP6:%.*]] = call i64 @llvm.vscale.i64()
+; IF-EVL-NEXT:    [[TMP7:%.*]] = mul i64 [[TMP6]], 4
+; IF-EVL-NEXT:    [[TMP8:%.*]] = sub i64 [[TMP7]], 1
+; IF-EVL-NEXT:    [[N_RND_UP:%.*]] = add i64 [[N]], [[TMP8]]
+; IF-EVL-NEXT:    [[N_MOD_VF:%.*]] = urem i64 [[N_RND_UP]], [[TMP5]]
+; IF-EVL-NEXT:    [[N_VEC:%.*]] = sub i64 [[N_RND_UP]], [[N_MOD_VF]]
+; IF-EVL-NEXT:    [[TRIP_COUNT_MINUS_1:%.*]] = sub i64 [[N]], 1
+; IF-EVL-NEXT:    [[TMP9:%.*]] = call i64 @llvm.vscale.i64()
+; IF-EVL-NEXT:    [[TMP10:%.*]] = mul i64 [[TMP9]], 4
+; IF-EVL-NEXT:    [[TMP11:%.*]] = insertelement <vscale x 4 x i32> zeroinitializer, i32 [[START:%.*]], i32 0
+; IF-EVL-NEXT:    [[BROADCAST_SPLATINSERT1:%.*]] = insertelement <vscale x 4 x i64> poison, i64 [[TRIP_COUNT_MINUS_1]], i64 0
+; IF-EVL-NEXT:    [[BROADCAST_SPLAT2:%.*]] = shufflevector <vscale x 4 x i64> [[BROADCAST_SPLATINSERT1]], <vscale x 4 x i64> poison, <vscale x 4 x i32> zeroinitializer
+; IF-EVL-NEXT:    br label [[VECTOR_BODY:%.*]]
+; IF-EVL:       vector.body:
+; IF-EVL-NEXT:    [[INDEX:%.*]] = phi i64 [ 0, [[VECTOR_PH]] ], [ [[INDEX_NEXT:%.*]], [[VECTOR_BODY]] ]
+; IF-EVL-NEXT:    [[EVL_BASED_IV:%.*]] = phi i64 [ 0, [[VECTOR_PH]] ], [ [[INDEX_EVL_NEXT:%.*]], [[VECTOR_BODY]] ]
+; IF-EVL-NEXT:    [[VEC_PHI:%.*]] = phi <vscale x 4 x i32> [ [[TMP11]], [[VECTOR_PH]] ], [ [[TMP20:%.*]], [[VECTOR_BODY]] ]
+; IF-EVL-NEXT:    [[TMP12:%.*]] = sub i64 [[N]], [[EVL_BASED_IV]]
+; IF-EVL-NEXT:    [[TMP13:%.*]] = call i32 @llvm.experimental.get.vector.length.i64(i64 [[TMP12]], i32 4, i1 true)
+; IF-EVL-NEXT:    [[TMP14:%.*]] = add i64 [[EVL_BASED_IV]], 0
+; IF-EVL-NEXT:    [[BROADCAST_SPLATINSERT:%.*]] = insertelement <vscale x 4 x i64> poison, i64 [[EVL_BASED_IV]], i64 0
+; IF-EVL-NEXT:    [[BROADCAST_SPLAT:%.*]] = shufflevector <vscale x 4 x i64> [[BROADCAST_SPLATINSERT]], <vscale x 4 x i64> poison, <vscale x 4 x i32> zeroinitializer
+; IF-EVL-NEXT:    [[TMP15:%.*]] = call <vscale x 4 x i64> @llvm.experimental.stepvector.nxv4i64()
+; IF-EVL-NEXT:    [[TMP16:%.*]] = add <vscale x 4 x i64> zeroinitializer, [[TMP15]]
+; IF-EVL-NEXT:    [[VEC_IV:%.*]] = add <vscale x 4 x i64> [[BROADCAST_SPLAT]], [[TMP16]]
+; IF-EVL-NEXT:    [[TMP17:%.*]] = icmp ule <vscale x 4 x i64> [[VEC_IV]], [[BROADCAST_SPLAT2]]
+; IF-EVL-NEXT:    [[TMP18:%.*]] = getelementptr inbounds i32, ptr [[A:%.*]], i64 [[TMP14]]
+; IF-EVL-NEXT:    [[TMP19:%.*]] = getelementptr inbounds i32, ptr [[TMP18]], i32 0
+; IF-EVL-NEXT:    [[VP_OP_LOAD:%.*]] = call <vscale x 4 x i32> @llvm.vp.load.nxv4i32.p0(ptr align 4 [[TMP19]], <vscale x 4 x i1> shufflevector (<vscale x 4 x i1> insertelement (<vscale x 4 x i1> poison, i1 true, i64 0), <vscale x 4 x i1> poison, <vscale x 4 x i32> zeroinitializer), i32 [[TMP13]])
+; IF-EVL-NEXT:    [[TMP20]] = add <vscale x 4 x i32> [[VP_OP_LOAD]], [[VEC_PHI]]
```

Comment:
Having another look at this after also taking a look at https://github.com/llvm/llvm-project/pull/93854.

Won't this poison some lanes of the reduction phi in the last (and possibly second-to-last) iteration, as the masked off lanes by the load are poisoned? If that's the case, then it might need to restriction to in-loop reductions, as the per-iteration reduction can make use of EVL of the current iteration.

If this is indeed an issue, I am curious if this would show up in runtime testing and if not it would be interesting how this could be improved?

Reply:
Assuming that it's impossible to get wraparound of IV + EVL, Mel's current solution does work right. However, if wraparound is possible such select is a problem and we do really need to emit `vp.merge` to switch to TU policy.
Also, such style of vectorization must avoid use of `ActiveLaneMask` at all costs and it has worse performance impact than properly using runtime VL

Generally, this PR is a good first step to emit vp intrinsics for reductions and follow up PR(s) will introduce `vp.merge` to eliminate use of `ActiveLaneMask`

Reply:
@fhahn  I do agree that as-is the changeset is generally unsafe for regular reduction case and the proper fix requires to add another recipe that emits `vp.merge` which later CG transforms into TU. But introduction of that new recipe will complicate the changeset, so how about add a clopt to control reduction vectorization which will be off by default ?  The follow up patch will add the missed recipe to make code correct.

Reply:
IIUC it should be safe for in-loop reductions, right? If that's the case, how about restricting codegen to in-loop reductions for now? this still allows brining up most of the code without miscompiles or the need for a temporary flag.

I am curious if at the moment it would be feasible to catch this issue at runtime, i.e. is there HW that implements EVL in a way that performs the last 2 iterations < VF?

Reply:
I thought it again and realized that we might not need https://github.com/llvm/llvm-project/pull/89580#discussion_r1580754264. The issue probably lies with the mask. I believe 67b3621dbcc8621a2bf9cad265d4213b3018b383 should solve the problem with the second-to-last iteration. The next patch will be performance improvement—transforming the select into a recipe that can generate vp.merge.

Reply:
In-loop reduction might also be unsafe. Please refer to `@cond_add_pred` in `llvm/test/Transforms/LoopVectorize/RISCV/vectorize-force-tail-with-evl-cond-reduction.ll` ([shortcut link](https://github.com/llvm/llvm-project/pull/90184/commits/67b3621dbcc8621a2bf9cad265d4213b3018b383#diff-bb6adf934f6e891dee24590af066f2fb6c5a102051216da8b06aa7f89aa642cfL373)).
I will separate out 67b3621 and open a new pull request.

Reply:
Opened https://github.com/llvm/llvm-project/pull/95368 for the issue of the second-to-last iteration.

Reply:
Ah I see. In that case, to avoid needing to pull in another patch, can we bail out for any operation that use something other than the header mask, so we can land this change, followed by https://github.com/llvm/llvm-project/pull/95368 ?

Reply:
I discussed it with @nikolaypanchenko further and realized my oversight.

The conclusion is that, yes, in-loop reduction is correct and safe, even for examples like `@cond_add_pred`. Previously, I was fixated on the mask needing to be completely correct. However, we only need to ensure that the mask is correct from `lane 0` to `lane evl - 1`. 
Therefore, I will limit this PR to in-loop reduction. Support for out-loop reduction will follow once `VPInstructions::Select (header_mask, true_value, false_value)` can be transformed into `vp.merge`.

Reply:
Sounds good, that matches my understanding as well